ICSharpCode.Decompiler/IL/Transforms/LocalFunctionDecompiler.cs

```diff
@@ -1,20 +1,314 @@
+// without restriction, including without limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons
+// to whom the Software is furnished to do so, subject to the following conditions:
+// 
+// The above copyright notice and this permission notice shall be included in all copies or
+// substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+// PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+// FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
 using System.Reflection;
 using System.Reflection.Metadata;
 using System.Text;
 using System.Text.RegularExpressions;
+using ICSharpCode.Decompiler.CSharp;
 using ICSharpCode.Decompiler.Metadata;
+using ICSharpCode.Decompiler.TypeSystem;
+using ICSharpCode.Decompiler.TypeSystem.Implementation;
 using ICSharpCode.Decompiler.Util;
 
 namespace ICSharpCode.Decompiler.IL.Transforms
 {
+	/// <summary>
+	/// Decompiler step for C# 7.0 local functions
+	/// </summary>
 	class LocalFunctionDecompiler : IILTransform
 	{
+		ILTransformContext context;
+
+		/// <summary>
+		/// The transform works like this:
+		/// 
+		/// <para>
+		/// local functions can either be used in method calls, i.e., call and callvirt instructions,
+		/// or can be used as part of the "delegate construction" pattern, i.e., <c>newobj Delegate(&lt;target-expression&gt;, ldftn &lt;method&gt;)</c>.
+		/// </para>
+		/// As local functions can be declared practically anywhere, we have to take a look at all use-sites and infer the declaration location from that. Use-sites can be call, callvirt and ldftn instructions.
+		/// After all use-sites are collected we construct the ILAst of the local function and add it to the parent function.
+		/// Then all use-sites of the local-function are transformed to a call to the <c>LocalFunctionMethod</c> or a ldftn of the <c>LocalFunctionMethod</c>.
+		/// In a next step we handle all nested local functions.
+		/// After all local functions are transformed, we move all local functions that do not capture any variables to the top-level function.
+		/// </summary>
 		public void Run(ILFunction function, ILTransformContext context)
 		{
-			throw new NotImplementedException();
+			if (!context.Settings.LocalFunctions)
+				return;
+			this.context = context;
+			var localFunctions = new Dictionary<IMethod, List<CallInstruction>>();
+			var cancellationToken = context.CancellationToken;
+			// Find use-sites
```

Comment:
As we discussed on Skype today, this needs a re-work to reliably handle mutually recursive local functions.